# <center><font color='MediumSeaGreen'>Desafío - Regresión desde el aprendizaje de máquinas </font><center>

### <center><b> Bárbara Pérez </b><center>

_________

## <font color='MediumSeaGreen'>Descripción</font>

Una consultora internacional radicada en EEUU está buscando analistas, por lo que han diseñado una prueba que permita poder seleccionar a quienes cuenten con las habilidades necesarias para crear modelación estadística. Para ello, ponen a disposición de los interesados una base de datos sobre los precios de las viviendas en Boston, utilizada en el paper Harrison Jr, D., & Rubinfeld, D. L. (1978). Hedonic housing prices and the demand for clean air. Journal of environmental economics and management, 5(1), 81-102.

El objetivo del ejercicio de captación de talento es desarrollar un modelo predictivo para el valor mediano de las casas mediante el entrenamiento de un modelo de regresión lineal.


● <span style='background:gainsboro'><font color='dimgray'>crim</font></span> : Tasa de criminalidad por sector de Boston.

● <span style='background:gainsboro'><font color='dimgray'>zn</font></span> :proporción de terreno residencial asignado para terrenos baldíos.

● <span style='background:gainsboro'><font color='dimgray'>indus</font></span>: proporción de negocios no asociados al comercio por sector.

● <span style='background:gainsboro'><font color='dimgray'>chas</font></span>: Dummy. 1 si el sector colinda con el río Charles, 0 de lo contrario.

● <span style='background:gainsboro'><font color='dimgray'>nox</font></span>: Concentración de dióxido de carbono.

● <span style='background:gainsboro'><font color='dimgray'>rm</font></span>: cantidad promedio de habitaciones por casa.

● <span style='background:gainsboro'><font color='dimgray'>age</font></span>: proporción de casas construidas antes de 1940.

● <span style='background:gainsboro'><font color='dimgray'>dis</font></span> : distancia promedio a cinco centros de empleos.

● <span style='background:gainsboro'><font color='dimgray'>rad</font></span>: índice de accesibilidad a autopistas.

● <span style='background:gainsboro'><font color='dimgray'>tax</font></span>: nivel de impuestos asociados a viviendas.

● <span style='background:gainsboro'><font color='dimgray'>ptratio</font></span>: razón alumno:profesor por sector de Boston.

● <span style='background:gainsboro'><font color='dimgray'>black</font></span>: proporción de afroamericanos por sector de Boston.

● <span style='background:gainsboro'><font color='dimgray'>lstat</font></span>: porcentaje de población de estratos bajos.

● <span style='background:gainsboro'><font color='dimgray'>medv</font></span>: valor mediano de las casas.

____________

## <font color='MediumSeaGreen'>Requerimientos</font> 

A continuación revisaremos los requerimientos y acciones que la empresa a la cual postulas te pide realizar.

<b><font color='MediumSeaGreen'>1.</font> Preparar el ambiente de trabajo (1 puntos)</b>

● Importe las librerías básicas para el análisis de datos.

● Importe el módulo *linear_model* , y las funciones *mean_squared_error* , *r2_score* y *train_test_split* .

● Importe la base de datos *boston.csv* y elimine la columna *Unnamed: 0* .
    
● Obtenga las medidas descriptivas de la base de datos con *.describe()* .


In [1]:
#Importar librerias 
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import scipy.stats as stats
import statsmodels.formula.api as smf
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
#Importar módulo linear_model , y las funciones mean_squared_error , r2_score y train_test_split .

from sklearn import linear_model
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split

In [3]:
#Importar dataframe
df= pd.read_csv('boston.csv')

In [4]:
#Reviso el dataframe
df.head()

Unnamed: 0     crim    zn  indus  chas    nox     rm   age     dis  rad  \
0           1  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1   
1           2  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2   
2           3  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2   
3           4  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3   
4           5  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3   

   tax  ptratio   black  lstat  medv  
0  296     15.3  396.90   4.98  24.0  
1  242     17.8  396.90   9.14  21.6  
2  242     17.8  392.83   4.03  34.7  
3  222     18.7  394.63   2.94  33.4  
4  222     18.7  396.90   5.33  36.2

In [5]:
#elimine la columna Unnamed: 0 .
df = df.drop(columns='Unnamed: 0', axis=1)

In [6]:
#Revisamos los resultados de eliminar Unnamed: 0
df.head()

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

    black  lstat  medv  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2

In [7]:
#Obtenga las medidas descriptivas de la base de datos con .describe() 
df.describe()

crim          zn       indus        chas         nox          rm  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              age         dis         rad         tax     ptratio       black  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            lstat        medv  
count  506.000000  506.000000  
mean    12.653063   22.532806  
std      7.141062    9.197104  
min      1.730000    5.000000  
25%      6.950000   17.025000  
50%     11.360000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

____

<b><font color='MediumSeaGreen'>2.</font> Dividir la muestra (1 puntos)</b>


● Genere conjuntos de entrenamiento y pruebas con *train_test_split* .

● Reserve un 33% de la muestra para el conjunto de pruebas.

● Incluya una semilla pseudoaleatoria a su elección, esto lo puede hacer con el
argumento *random_state* dentro del método *train_test_plit*.

In [8]:
#Se realiza con la variable medv porque es el que tiene un mayor mean
df_var_y= df['medv']
df_var_x= df.drop(columns='medv')

In [9]:
#SE seleccionará un conjunto de datos para entrenamiento y validacion, 
#Tambien se colocoará la semilla
x_train, x_test, y_train, y_test = train_test_split(df_var_x, df_var_y, test_size=0.33, random_state=6)

In [10]:
type(y_test)

pandas.core.series.Series

_______

<b><font color='MediumSeaGreen'>3.</font> Generar modelos (2 puntos)</b>


● Ahora implementaremos dos versiones del modelo lineal:

    ○ Con intercepto.
    
    ○ Sin intercepto.
    
● Cada versión debe generarse en un nuevo objeto inicializado.

● Posteriormente se deben entrenar los modelos especificando la matriz y vector
de entrenamiento.

● Con los modelos entrenados, genere una predicción de la matriz de pruebas con
el método *.predict()*.

In [11]:
modelo_con_intercepto = linear_model.LinearRegression(fit_intercept=True)
modelo_con_intercepto.fit(x_train, y_train) #ingresamos el training set
modelo_con_intercepto_yhat = modelo_con_intercepto.predict(x_test) #Para medir el desempeño predictivo del modelo es generar nuevas predicciones en nuestra muestra de prueba

In [12]:
modelo_sin_intercepto = linear_model.LinearRegression(fit_intercept=False)
modelo_sin_intercepto.fit(x_train, y_train) #ingresamos el training set
modelo_sin_intercepto_yhat = modelo_sin_intercepto.predict(x_test) #Para medir el desempeño predictivo del modelo es generar nuevas predicciones en nuestra muestra de prueba

______

<b><font color='MediumSeaGreen'>4.</font> Obtención de métricas (1 puntos)</b>

● Ahora generaremos una función llamada *report_scores* que ingrese como
argumentos el vector de datos predichos y el vector de datos por validar.

● La función debe imprimir las métricas del Error Cuadrático Promedio y R2.

● Reporte las métricas para ambos modelos. En base a ello, seleccione el mejor
modelo.

In [13]:
def report_scores (modelo_yhat, y_test):
    modelo_mse = mean_squared_error(y_test, modelo_yhat).round(2)
    modelo_r2 = r2_score(y_test, modelo_yhat).round(2)
    print(f"Mean Squared Error (mse): {modelo_mse}")
    print(f"R-cuadrado: {modelo_r2}")

In [14]:
#Modelo con intercepto
report_scores (modelo_con_intercepto_yhat, y_test)

Mean Squared Error (mse): 26.62
R-cuadrado: 0.7


In [15]:
#Modelo sin itnercepto
report_scores (modelo_sin_intercepto_yhat, y_test)

Mean Squared Error (mse): 28.69
R-cuadrado: 0.68


Comparando ambos modelos, el modelo con intercepto presenta un mayor R cuadrado y un menor error cuadratico (mse)

_________

<b><font color='MediumSeaGreen'>5.</font> Refactorización del modelo (1 puntos)</b>

● Genere una función llamada *fetch_features* que ingrese como argumentos la
base de datos y el nombre del vector objetivo. El nombre del vector debe ser
<span style='background:gainsboro'><font color='dimgray'>medv</font></span> por defecto.

● La función debe retornar una lista con las correlaciones entre cada atributo y el vector objetivo y su nombre.

● Reporte brevemente cuales son los 6 atributos con una mayor correlación
absoluta con <span style='background:gainsboro'><font color='dimgray'>medv</font></span> (de mayor a menor correlación).

In [16]:
def fetch_features (df, vector_objetivo):
    # extraemos los nombres de las columnas en la base de datos
    columns = list(df.columns)
    # generamos 3 arrays vacíos para guardar los valores
    # nombre de la variable, # correlación de pearson, #valor absoluto de la correlación
    attr_name, pearson_r, abs_pearson_r = [], [], []
    # para cada columna en el array de columnas
    for col in columns:
    # si la columna no es la dependiente
        if col != vector_objetivo:
        # adjuntar el nombre de la variable en attr_name
            attr_name.append(col)
            # adjuntar la correlación de pearson
            pearson_r.append(df[col].corr(df[vector_objetivo]))
            # adjuntar el absoluto de la correlación de pearson
            abs_pearson_r.append(abs(df[col].corr(df[vector_objetivo])))
        
    # transformamos los arrays en un DataFrame
    features = pd.DataFrame({
    'attribute': attr_name,
    'corr':pearson_r,
    'abs_corr':abs_pearson_r
    })
    
    # generamos el index con los nombres de las variables
    features = features.set_index('attribute')
    # ordenamos los valores de forma descendiente
    features.sort_values('abs_corr', ascending=False)
    
    return features

In [17]:
#Obtenermos el df con las correlaciones que tenga medv por defecto
correlaciones_medv = fetch_features(df, 'medv')

**Reporte brevemente cuales son los 6 atributos con una mayor correlación absoluta con medv (de mayor a menor correlación).**

In [18]:
correlaciones_medv.head(6)

corr  abs_corr
attribute                    
crim      -0.388305  0.388305
zn         0.360445  0.360445
indus     -0.483725  0.483725
chas       0.175260  0.175260
nox       -0.427321  0.427321
rm         0.695360  0.695360

In [19]:
correlaciones_medv.sort_values('abs_corr',ascending=False).head(6)

corr  abs_corr
attribute                    
lstat     -0.737663  0.737663
rm         0.695360  0.695360
ptratio   -0.507787  0.507787
indus     -0.483725  0.483725
tax       -0.468536  0.468536
nox       -0.427321  0.427321

Se parecia que los atributos con mayor correlación absoluta con 'medv'son:

    - lstat
    - rm
    - ptratio
    - indus
    - tax
    - nox

_________

<b><font color='MediumSeaGreen'>6.</font> Refactorización del modelo predictivo (2 puntos)</b>

● Genere otros conjuntos de entrenamiento y validación en base a una matriz con
los 6 atributos identificados en el ejercicio anterior, y el vector objetivo.

● Entrene un modelo en base al mejor desempeño.

● Reporte las métricas para el nuevo modelo.

In [20]:
#De los datos obtenidos anteriormente generamos una lista con sus index para obtener sus nombres
atributos_6 = list(correlaciones_medv.sort_values('abs_corr',ascending=False).head(6).index)

In [21]:
atributos_6

['lstat', 'rm', 'ptratio', 'indus', 'tax', 'nox']

In [22]:
#Se crea un subset con los 6 atributos 
df_var_x_6= df.loc[:, atributos_6]

In [23]:
#Se vuelve a generar nuevo conjutnop de entrenamiento y validacion para los datos entregados
x_train_2, x_test_2, y_train_2, y_test_2 = train_test_split(df_var_x_6, df_var_y, test_size=.33,
random_state=27)

In [24]:
modelo_refact_con_inter = linear_model.LinearRegression(fit_intercept=True)
modelo_refact_con_inter.fit(x_train_2, y_train_2) #ingresamos el training set
modelo_refact_con_inter_yhat = modelo_refact_con_inter.predict(x_test_2) #Para medir el desempeño predictivo del modelo es generar nuevas predicciones en nuestra muestra de prueba

In [25]:
modelo_refact_sin_inter = linear_model.LinearRegression(fit_intercept=False)
modelo_refact_sin_inter.fit(x_train_2, y_train_2) #ingresamos el training set
modelo_refact_sin_inter_yhat = modelo_refact_sin_inter.predict(x_test_2) #Para medir el desempeño predictivo del modelo es generar nuevas predicciones en nuestra muestra de prueba

In [26]:
#Corremos el modelo refactorizado con intercepto
report_scores(modelo_refact_con_inter_yhat, y_test_2)

Mean Squared Error (mse): 29.8
R-cuadrado: 0.66


In [27]:
#Corremos el modelo refactorizado sin intercepto
report_scores(modelo_refact_sin_inter_yhat, y_test_2)

Mean Squared Error (mse): 31.71
R-cuadrado: 0.64


In [28]:
#Corremos el mejor modelo anteriormente (con intercepto)
#Modelo con intercepto
report_scores (modelo_con_intercepto_yhat, y_test)

Mean Squared Error (mse): 26.62
R-cuadrado: 0.7


Comparando los modelos refactorizados el que presenta un mejor resultado es el modelo con intercepto, presentando un R cuadrado de 0.66 y un mse menor.

Sin embargo, al comparar el modelo refactorizado vs el mejor sin refactorizar se observa que entrega mejores resultadoes el modelo sin refactorizar con intercepto, teniendo un R cuadrado mayor y un menor error cuadratico (mse).

_____

<b><font color='MediumSeaGreen'>7.</font> Predicción de casos (2 puntos)</b>: 
    A continuación se generaron dos arrays que
representan el peor escenario posible (*worst_neighbor*) y el mejor escenario
posible (*best_neighbor*). Las variables representan, para cada caso, los valores de
los siguientes atributos (en el mismo orden entregado): **'lstat', 'rm',
'ptratio', 'indus', 'tax', 'nox'.**

<span style='background:gainsboro'><font color='dimgray'>worst_neighbor = np.array([37.9, 12.6, 3.5, 27.7, 187, 0.87]).reshape(1,
-1)</font></span>
    
    
<span style='background:gainsboro'><font color='dimgray'>best_neighbor = np.array([1.73, 22, 8.7, 0.46, 711, 0.38]).reshape(1,
-1)</font></span>

● Ingrese los arrays en el modelo entrenado en el ejercicio anterior, y reporte la predicción entregada por el modelo.

In [29]:
worst_neighbor = np.array([37.9, 12.6, 3.5, 27.7, 187, 0.87]).reshape(1, -1)
best_neighbor = np.array([1.73, 22, 8.7, 0.46, 711, 0.38]).reshape(1,-1)

In [30]:
modelo_refact_con_inter.predict(worst_neighbor).round(2)

C:\ProgramData\Anaconda3\envs\rstudio\lib\site-packages\sklearn\base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


array([57.98])

**El valor esperado para worst_neighbor con:
el modelo refactorizado con intercepto es de 57.98**

In [31]:
modelo_refact_sin_inter.predict(worst_neighbor).round(2)

C:\ProgramData\Anaconda3\envs\rstudio\lib\site-packages\sklearn\base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


array([62.05])

**El valor esperado para worst_neighbor con:
el modelo refactorizado sin intercepto es de 62.05**

In [32]:
modelo_refact_con_inter.predict(best_neighbor).round(2)

C:\ProgramData\Anaconda3\envs\rstudio\lib\site-packages\sklearn\base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


array([117.91])

**El valor esperado para best_neighbor con:
el modelo refactorizado con intercepto es de 117.91**

In [33]:
modelo_refact_sin_inter.predict(best_neighbor).round(2)

C:\ProgramData\Anaconda3\envs\rstudio\lib\site-packages\sklearn\base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


array([128.57])

**El valor esperado para best_neighbor con:
el modelo refactorizado sin intercepto es de 128.57**

________